### ISSUE TITLE
Race on NettyServer shutdown.

### ISSUE THREAD

**Issue by carl-mastrangelo** [grpc/grpc-java#2014]
Even with 2292cbfca091a4c6bb315a291929c656a81b90b6 it seems there is still a race with shutdown.  Initial guess is that the shutdown and close are happening in the Boss Event loop and the normal event loop.  Conflicting reads and writes:

```
WARNING: ThreadSanitizer: data race (pid=705947)
  Read of size 8 at 0x7fa79fc81808 by thread T28 (mutexes: write M109229):
    #0 io.netty.channel.AbstractChannel$AbstractUnsafe.outboundBuffer()Lio/netty/channel/ChannelOutboundBuffer; (AbstractChannel.java:434)  
    #1 io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.init(Lio/netty/channel/AbstractChannelHandlerContext$AbstractWriteTask;Lio/netty/channel/AbstractChannelHandlerContext;Ljava/lang/Object;Lio/netty/channel/ChannelPromise;)V (AbstractChannelHandlerContext.java:1027)  
    #2 io.netty.channel.AbstractChannelHandlerContext$WriteAndFlushTask.newInstance(Lio/netty/channel/AbstractChannelHandlerContext;Ljava/lang/Object;Lio/netty/channel/ChannelPromise;)Lio/netty/channel/AbstractChannelHandlerContext$WriteAndFlushTask; (AbstractChannelHandlerContext.java:1101)  
    #3 io.netty.channel.AbstractChannelHandlerContext$WriteAndFlushTask.access$1600(Lio/netty/channel/AbstractChannelHandlerContext;Ljava/lang/Object;Lio/netty/channel/ChannelPromise;)Lio/netty/channel/AbstractChannelHandlerContext$WriteAndFlushTask; (AbstractChannelHandlerContext.java:1089)  
    #4 io.netty.channel.AbstractChannelHandlerContext.write(Ljava/lang/Object;ZLio/netty/channel/ChannelPromise;)V (AbstractChannelHandlerContext.java:807)  
    #5 io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(Ljava/lang/Object;Lio/netty/channel/ChannelPromise;)Lio/netty/channel/ChannelFuture; (AbstractChannelHandlerContext.java:780)  
    #6 io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(Ljava/lang/Object;)Lio/netty/channel/ChannelFuture; (AbstractChannelHandlerContext.java:817)  
    #7 io.netty.channel.DefaultChannelPipeline.writeAndFlush(Ljava/lang/Object;)Lio/netty/channel/ChannelFuture; (DefaultChannelPipeline.java:1011)  
    #8 io.netty.channel.AbstractChannel.writeAndFlush(Ljava/lang/Object;)Lio/netty/channel/ChannelFuture; (AbstractChannel.java:289)  
    #9 io.grpc.netty.NettyServerTransport.shutdownNow(Lio/grpc/Status;)V (NettyServerTransport.java:101)  
    #10 io.grpc.internal.ServerImpl$ServerListenerImpl.serverShutdown()V (ServerImpl.java:302)  
    #11 io.grpc.netty.NettyServer$2.operationComplete(Lio/netty/channel/ChannelFuture;)V (NettyServer.java:178)  
    #12 io.grpc.netty.NettyServer$2.operationComplete(Lio/netty/util/concurrent/Future;)V (NettyServer.java:171)  
    #13 io.netty.util.concurrent.DefaultPromise.notifyListener0(Lio/netty/util/concurrent/Future;Lio/netty/util/concurrent/GenericFutureListener;)V (DefaultPromise.java:518)  
    #14 io.netty.util.concurrent.DefaultPromise.notifyListenersNow()V (DefaultPromise.java:492)  
    #15 io.netty.util.concurrent.DefaultPromise.notifyListenersWithStackOverFlowProtection()V (DefaultPromise.java:431)  
    #16 io.netty.util.concurrent.DefaultPromise.notifyListeners()V (DefaultPromise.java:420)  
    #17 io.netty.util.concurrent.DefaultPromise.trySuccess(Ljava/lang/Object;)Z (DefaultPromise.java:108)  
    #18 io.netty.channel.DefaultChannelPromise.trySuccess()Z (DefaultChannelPromise.java:82)  
    #19 io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetSuccess(Lio/netty/channel/ChannelPromise;)V (AbstractChannel.java:884)  
    #20 io.netty.channel.AbstractChannel$AbstractUnsafe.doClose0(Lio/netty/channel/ChannelPromise;)V (AbstractChannel.java:675)  
    #21 io.netty.channel.AbstractChannel$AbstractUnsafe.close(Lio/netty/channel/ChannelPromise;Ljava/lang/Throwable;Z)V (AbstractChannel.java:652)  
    #22 io.netty.channel.AbstractChannel$AbstractUnsafe.close(Lio/netty/channel/ChannelPromise;)V (AbstractChannel.java:596)  
    #23 io.netty.channel.DefaultChannelPipeline$HeadContext.close(Lio/netty/channel/ChannelHandlerContext;Lio/netty/channel/ChannelPromise;)V (DefaultChannelPipeline.java:1256)  
    #24 io.netty.channel.AbstractChannelHandlerContext.invokeClose(Lio/netty/channel/ChannelPromise;)V (AbstractChannelHandlerContext.java:610)  
    #25 io.netty.channel.AbstractChannelHandlerContext.access$1100(Lio/netty/channel/AbstractChannelHandlerContext;Lio/netty/channel/ChannelPromise;)V (AbstractChannelHandlerContext.java:36)  
    #26 io.netty.channel.AbstractChannelHandlerContext$13.run()V (AbstractChannelHandlerContext.java:599)  
    #27 io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(J)Z (SingleThreadEventExecutor.java:339)  
    #28 io.netty.channel.nio.NioEventLoop.run()V (NioEventLoop.java:393)  
    #29 io.netty.util.concurrent.SingleThreadEventExecutor$5.run()V (SingleThreadEventExecutor.java:742)  
    #30 io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run()V (DefaultThreadFactory.java:145)  
    #31 java.lang.Thread.run()V (Thread.java:745)  
    #32 (Generated Stub)  

  Previous write of size 8 at 0x7fa79fc81808 by thread T65:
    #0 io.netty.channel.AbstractChannel$AbstractUnsafe.close(Lio/netty/channel/ChannelPromise;Ljava/lang/Throwable;Z)V (AbstractChannel.java:626)  
    #1 io.netty.channel.AbstractChannel$AbstractUnsafe.close(Lio/netty/channel/ChannelPromise;)V (AbstractChannel.java:596)  
    #2 io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.closeOnRead(Lio/netty/channel/ChannelPipeline;)V (AbstractNioByteChannel.java:76)  
    #3 io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read()V (AbstractNioByteChannel.java:131)  
    #4 io.netty.channel.nio.NioEventLoop.processSelectedKey(Ljava/nio/channels/SelectionKey;Lio/netty/channel/nio/AbstractNioChannel;)V (NioEventLoop.java:563)  
    #5 io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized([Ljava/nio/channels/SelectionKey;)V (NioEventLoop.java:504)  
    #6 io.netty.channel.nio.NioEventLoop.processSelectedKeys()V (NioEventLoop.java:418)  
    #7 io.netty.channel.nio.NioEventLoop.run()V (NioEventLoop.java:390)  
    #8 io.netty.util.concurrent.SingleThreadEventExecutor$5.run()V (SingleThreadEventExecutor.java:742)  
    #9 io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run()V (DefaultThreadFactory.java:145)  
    #10 java.lang.Thread.run()V (Thread.java:745)  
    #11 (Generated Stub)  

```

**Comment by carl-mastrangelo** [comment]
Interestingly, the one racing thread created the other.   Perhaps Unsafe is just not being safely accessed?

Here are the stacks of Thread creation:

```
  Thread T65 (tid=706607, running) created by thread T28 at:
    #3 io.netty.util.concurrent.ThreadPerTaskExecutor.execute(Ljava/lang/Runnable;)V (ThreadPerTaskExecutor.java:33)  
    #4 io.netty.util.concurrent.SingleThreadEventExecutor.doStartThread()V (SingleThreadEventExecutor.java:731)  
    #5 io.netty.util.concurrent.SingleThreadEventExecutor.startThread()V (SingleThreadEventExecutor.java:724)  
    #6 io.netty.util.concurrent.SingleThreadEventExecutor.execute(Ljava/lang/Runnable;)V (SingleThreadEventExecutor.java:671)  
    #7 io.netty.channel.AbstractChannel$AbstractUnsafe.register(Lio/netty/channel/EventLoop;Lio/netty/channel/ChannelPromise;)V (AbstractChannel.java:468)  
    #8 io.netty.channel.SingleThreadEventLoop.register(Lio/netty/channel/ChannelPromise;)Lio/netty/channel/ChannelFuture; (SingleThreadEventLoop.java:56)  
    #9 io.netty.channel.SingleThreadEventLoop.register(Lio/netty/channel/Channel;)Lio/netty/channel/ChannelFuture; (SingleThreadEventLoop.java:50)  
    #10 io.netty.channel.MultithreadEventLoopGroup.register(Lio/netty/channel/Channel;)Lio/netty/channel/ChannelFuture; (MultithreadEventLoopGroup.java:85)  
    #11 io.netty.bootstrap.ServerBootstrap$ServerBootstrapAcceptor.channelRead(Lio/netty/channel/ChannelHandlerContext;Ljava/lang/Object;)V (ServerBootstrap.java:245)  
    #12 io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(Ljava/lang/Object;)V (AbstractChannelHandlerContext.java:348)  
    #13 io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(Lio/netty/channel/AbstractChannelHandlerContext;Ljava/lang/Object;)V (AbstractChannelHandlerContext.java:334)  
    #14 io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(Ljava/lang/Object;)Lio/netty/channel/ChannelHandlerContext; (AbstractChannelHandlerContext.java:326)  
    #15 io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(Lio/netty/channel/ChannelHandlerContext;Ljava/lang/Object;)V (DefaultChannelPipeline.java:1320)  
    #16 io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(Ljava/lang/Object;)V (AbstractChannelHandlerContext.java:348)  
    #17 io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(Lio/netty/channel/AbstractChannelHandlerContext;Ljava/lang/Object;)V (AbstractChannelHandlerContext.java:334)  
    #18 io.netty.channel.DefaultChannelPipeline.fireChannelRead(Ljava/lang/Object;)Lio/netty/channel/ChannelPipeline; (DefaultChannelPipeline.java:905)  
    #19 io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read()V (AbstractNioMessageChannel.java:93)  
    #20 io.netty.channel.nio.NioEventLoop.processSelectedKey(Ljava/nio/channels/SelectionKey;Lio/netty/channel/nio/AbstractNioChannel;)V (NioEventLoop.java:563)  
    #21 io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized([Ljava/nio/channels/SelectionKey;)V (NioEventLoop.java:504)  
    #22 io.netty.channel.nio.NioEventLoop.processSelectedKeys()V (NioEventLoop.java:418)  
    #23 io.netty.channel.nio.NioEventLoop.run()V (NioEventLoop.java:390)  
    #24 io.netty.util.concurrent.SingleThreadEventExecutor$5.run()V (SingleThreadEventExecutor.java:742)  
    #25 io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run()V (DefaultThreadFactory.java:145)  
    #26 java.lang.Thread.run()V (Thread.java:745)  
    #27 (Generated Stub)  
    
    
     Thread T28 (tid=706440, running) created by thread T5 at:
    #2 java.lang.Thread.start()V (Thread.java:714)  
    #3 io.netty.util.concurrent.ThreadPerTaskExecutor.execute(Ljava/lang/Runnable;)V (ThreadPerTaskExecutor.java:33)  
    #4 io.netty.util.concurrent.SingleThreadEventExecutor.doStartThread()V (SingleThreadEventExecutor.java:731)  
    #5 io.netty.util.concurrent.SingleThreadEventExecutor.startThread()V (SingleThreadEventExecutor.java:724)  
    #6 io.netty.util.concurrent.SingleThreadEventExecutor.execute(Ljava/lang/Runnable;)V (SingleThreadEventExecutor.java:671)  
    #7 io.netty.channel.AbstractChannel$AbstractUnsafe.register(Lio/netty/channel/EventLoop;Lio/netty/channel/ChannelPromise;)V (AbstractChannel.java:468)  
    #8 io.netty.channel.SingleThreadEventLoop.register(Lio/netty/channel/ChannelPromise;)Lio/netty/channel/ChannelFuture; (SingleThreadEventLoop.java:56)  
    #9 io.netty.channel.SingleThreadEventLoop.register(Lio/netty/channel/Channel;)Lio/netty/channel/ChannelFuture; (SingleThreadEventLoop.java:50)  
    #10 io.netty.channel.MultithreadEventLoopGroup.register(Lio/netty/channel/Channel;)Lio/netty/channel/ChannelFuture; (MultithreadEventLoopGroup.java:85)  
    #11 io.netty.bootstrap.AbstractBootstrap.initAndRegister()Lio/netty/channel/ChannelFuture; (AbstractBootstrap.java:327)  
    #12 io.netty.bootstrap.AbstractBootstrap.doBind(Ljava/net/SocketAddress;)Lio/netty/channel/ChannelFuture; (AbstractBootstrap.java:282)  
    #13 io.netty.bootstrap.AbstractBootstrap.bind(Ljava/net/SocketAddress;)Lio/netty/channel/ChannelFuture; (AbstractBootstrap.java:278)  
    #14 io.grpc.netty.NettyServer.start(Lio/grpc/internal/ServerListener;)V (NettyServer.java:152)  
    #15 io.grpc.internal.ServerImpl.start()Lio/grpc/internal/ServerImpl; (ServerImpl.java:144)  
    #16 io.grpc.internal.ServerImpl.start()Lio/grpc/Server; (ServerImpl.java:77)  
    #17 io.grpc.testing.integration.CompressionTest.compression()V (CompressionTest.java:209)
```

**Comment by buchgr** [comment]
the report is legit. will fix it in Netty.

**Comment by buchgr** [comment]
fixed in next netty release.

**Comment by buchgr** [comment]
fixed by 0ad5948
